[2단계 - 경로 조회 기능] 후추(주찬민) 미션 제출합니다.

Combi153 opened the pull request:
안녕하세요 조앤!

2단계 미션을 뒤늦게 제출합니다!

3단계까지 도전하고 싶다고 의사를 밝혔었는데, 아쉽게도 이번에는 2단계에서 마무리 지어야 할 것 같습니다 벌써 월요일이네요 😢 

하지만 2단계에서 배운 점도 많았습니다. 😄 Repository를 만들어 본 것은 처음이었는데 Repository를 도입하기 전 후의 코드를 보면서, 존재 이유를 조금은 알 것 같았습니다. 조앤의 리뷰로 인해 Repository 에 대한 이해가 더 늘어났습니다!

지난 리뷰에서 많이 배웠습니다!

이번 리뷰도 잘 부탁드립니다!

## 궁금한 점
### DTO
![image](https://github.com/woowacourse/jwp-subway-path/assets/106813090/cd815042-540a-4846-85ec-9571be2b1d07)

기능이 많아지면서 DTO의 개수가 많아집니다. 한 두 개 있을 때는 DTO를 관리하는 것이 어렵지 않았는데, 점차 어떤 상황에 어떤 DTO를 써야할 지 DTO를 직접 만든 저도 헷갈리더라구요. DTO를 관리하는 특별한 방법이 있으실까요? 가령 패키지 구조는 어떻게 가져가야 더 효율적일 수 있을지 궁금합니다.

Combi153 committed:
refactor: Sections 거리 정보 검증 로직 수정

Combi153 committed:
build: jgrapht 라이브러리 추가

Combi153 committed:
docs: 구현 기능 목록 추가

Combi153 committed:
test: jgrapht 학습 테스트 진행

Combi153 committed:
refactor: 패키지 구조 변경

Combi153 committed:
feat: jgrapht 학습 테스트 진행

Combi153 committed:
feat: SubwayMap 기능 추가

Combi153 committed:
feat: PathFinder 기능 추가

Combi153 committed:
feat: FarePolicy 기능 추가

Combi153 committed:
feat: PathService 기능 추가

Combi153 committed:
feat: PathController 기능 추가

Combi153 committed:
refactor: import 구문 수정

Combi153 committed:
refactor: @LocalServerPort 어노테이션 수정

Combi153 committed:
refactor: Sections 로직 수정

Combi153 committed:
refactor: LineRepository, SectionRepository 로직 수정

Combi153 committed:
refactor: StationRepository 로직 수정

Combi153 committed:
refactor: 패키지 구조 변경

Combi153 committed:
feat: Transactional 적용

Combi153 committed:
build: mysql 의존성 추가

Combi153 committed:
build: docker 설정

Combi153 committed:
build: 테스트, 프로덕션 DB 분리

Combi153 committed:
test: StationDao test 수정

Combi153 committed:
refactor: 노선 조회 API 수정

Combi153 committed:
test: LineServiceTest 검증 로직 추가

Combi153 committed:
docs: readme 작성

seovalue commented on README.md:
````diff
@@ -112,6 +185,31 @@ create table if not exists SECTION
+    );
+
+```
+## Docker
````
docker 관련을 추가해주시면서 불필요한 바이너리파일들이 같이 커밋되었어요. `.gitignore`에 명시해주시고 제거해주세요.

Combi153 replied:
넵 제거하고 .gitignore 에 관련 내용 추가했습니다!

seovalue commented on src/main/java/subway/dto/DistanceResponse.java:
```diff
@@ -0,0 +1,23 @@
+package subway.dto;
```
저는 보통 레이어별로 DTO를 나누는 편인데요.
예를 들자면 controller로 들어오고 나가는 요청/응답인 request,response는 controller 하위 패키지에 위치하도록 하고, dto라는 패키지 안에 request/response 패키지를 한번 더 분리해서 두는 편이에요.
서비스로 진입/그리고 나올 때 사용되는 dto는 service 하위에 두고요~~

Combi153 replied:
알려주셔서 감사합니다! 그렇다면 질문이 있는데요!

1. controller 에서 request값을 다시 서비스에서 사용하는 request로 재포장해서 service로 보내는 것인가요?
2. controller dto 에서 service dto 로 변환하는 로직이 필요할 것 같은데 그러한 로직은 어디에 두시는 편이신가요?

seovalue replied:
1. 바로 service로 넘길 때도 있고, 재포장하는 경우도 있어요! 하지만 두 경우 모두 시작점이 외부이기 때문에 controller/dto에 두는 편이에용
2. Controller에서 사용되는 DTO 내부에 controllerDto.toServiceDto() 와 같은 메서드를 만들어서 넘겨요! service에서 사용되는 dto는 나중에 또다른 모듈로 분리될 확률이 있다고 생각해서, 프레젠테이션 레이어가 코어를 알도록 하는 편이에요

seovalue commented on src/main/java/subway/dao/LineEntity.java:
```diff
@@ -18,12 +17,8 @@ public LineEntity(final Long id, final String name, final String color) {
+        return new LineEntity(line.getId(), line.getNameValue(), line.getColorValue());
     }
 
     public Line toLine(final List<Section> sections) {
```
entity가 도메인을 알고 있는 것에 대해서 어떻게 생각하시나요?

Combi153 replied:
두 가지 생각이 듭니다.

1. entity가 도메인을 알게 되어서 변환하기 편하다.
Repository는 Dao 에서 반환한 entity를 조합해 도메인으로 변환합니다. entity 에서 domain으로 변환하는 로직은 가독성이 떨어지는 로직입니다. 따라서 이를 entity에 두면 Repository의 가독성을 챙길 수 있습니다.

2. entity는 영속성 계층에 속하고 domain은 비즈니스 계층에 속합니다. entity가 도메인을 알게 되면 계층 간 분리가 모호해집니다. domain이 비즈니스 계층과 영속성 계층 모두에서 사용되기 때문입니다.

유지 보수를 용이하게 하기 위해 레이어드 아키텍쳐를 사용한 관점에서 entity가 도메인을 모르게 해야 할 것 같습니다. 관련 내용 리팩터링 해보겠습니다!

seovalue commented on src/main/java/subway/domain/section/Sections.java:
```diff
@@ -72,23 +75,21 @@ private boolean hasUniquely(final Station station) {
+        sections.add(changedSection);
+    }
+
+    private Optional<Section> getChangedSection(final Station from, final Station to, final int distance, final Direction direction) {
```
changedSection이 정확히 어떤걸 의미하는지 좀 더 나타내보면 어떨까요?

Combi153 replied:
넵 메서드의 동작에 대한 묘사를 더 자세히 할 수 있도록 이름을 수정하겠습니다!

seovalue commented on src/main/java/subway/domain/fare/BasicFarePolicy.java:
```diff
@@ -0,0 +1,28 @@
+
+import subway.domain.section.Distance;
+
+public class BasicFarePolicy implements FarePolicy {
```
FarePolicy라는 인터페이스를 두어 주신 이유가 궁금해요~ 구현체도 하나뿐이어서요ㅎㅎ
FirstDistanceFare, SecondDistanceFare, OverFare가 각각 존재하고 각 구현체 안에서 distance, increment 등을 관리한 뒤 calculate를 그에 맞게 해보면 어떨까요?

예를 들자면 BasicFare에는 basice_fare인 1250이 있고, distance_limit으로 10을 가져요.
그리고 nextFareChain에는 FirstDistanceFare를 갖구요.

그럼 calculate 안에서 아래처럼 동작하도록 할 수 있을 것 같아요!
```
@Override
    public int calculate(final Distance distance) {
        if (distance.isNotLongerThan(FIRST_DISTANCE)) {
            return BASIC_FARE;
        }
        return nextFareChain.calculate(distance);
    }
```

Combi153 replied:
FarePolicy 인터페이스를 둔 이유는 3단계에서 다른 요금 정책이 추가되는 것을 고려했기 때문입니다. 2단계에서는 거리에 따른 요금 정책만 존재하지만, 3단계에서는 노선별 추가 요금 정책이 있습니다. 이러한 정책을 FarePolicy 라는 인터페이스로 추상화해서 사용할 계획이었습니다!

거리 요금 정책에 대한 조언해주신 내용 좋은 것 같습니다. 거리 기준이나 increment 등의 변경 가능성에 대해 확장성을 갖춘 구조가 될 것 같습니다. 조언해주신 바 대로 리팩터링해보겠습니다!

다만, 요금 정책이 추가될 것을 대비해서 FarePolicy 인터페이스는 그대로 두고, DistanceFarePolicy로 FirstDistanceFarePolicy, SecondDistanceFarePolicy 등의 구현체를 추상화하겠습니다!

seovalue commented on src/main/java/subway/config/FareConfiguration.java:
```diff
@@ -0,0 +1,15 @@
+import subway.domain.fare.BasicFarePolicy;
+import subway.domain.fare.FarePolicy;
+
+@Configuration
```
존재 이유가 궁금해요

Combi153 replied:
도메인 객체가 스프링에 영향을 받지 않도록 코드를 짜고 싶었기 때문입니다!

PathService에 FarePolicy 필드를 추가하고 싶었습니다. 하지만 PathService는 스프링이 빈으로 관리해주는 @Service 어노테이션을 갖고 있습니다. 따라서 FarePolicy를 필드에 추가하기 위해 FarePolicy 역시 빈으로 관리될 필요가 있었습니다.

이때 FarePolicy는 도메인 객체였습니다. 도메인 객체는 최대한 스프링의 영향을 받지 않도록 코드를 작성하고 싶었습니다. 따라서 도메인 객체 상단에 @Component 등의 어노테이션을 달기보다, 외부에서 빈 등록이 될 수 있도록 설정을 해주었습니다.

seovalue commented on src/main/java/subway/application/PathService.java:
```diff
@@ -0,0 +1,61 @@
+        return new PathResponse(toResponses(path.getSectionEdges()), DistanceResponse.of(path.getDistance()), new FareResponse(fare));
+    }
+
+    private List<LineAndSectionsResponse> toResponses(final List<SectionEdge> sectionEdges) {
+        return sectionEdges.stream()
+                .collect(Collectors.groupingBy(SectionEdge::getLineId))
+                .entrySet().stream()
+                .map(entry -> {
+                    final Long lineId = entry.getKey();
+                    final Line line = lineRepository.findLineById(lineId);
+                    final List<Section> sections = entry.getValue().stream()
+                            .map(SectionEdge::toSection)
+                            .collect(Collectors.toUnmodifiableList());
+                    return LineAndSectionsResponse.of(line, sections);
+                }).collect(Collectors.toUnmodifiableList());
+    }
```
이런 부분은 `return LineAndSectionsResponse.from(sectionEdges);` 로 해서, LineAndSectionsResponse 클래스 내부에 구현되어있는 것은 어떨까요?

seovalue commented on src/main/java/subway/dto/PathResponse.java:
```diff
@@ -0,0 +1,31 @@
+
+public class PathResponse {
+
+    private List<LineAndSectionsResponse> lineAndSectionsResponse;
+    private DistanceResponse distanceResponse;
+    private FareResponse fareResponse;
+
```
이렇게 되면 응답의 json 형태가 항상 key를 distanceResponse: { "distance": 10 }, 이런 식으로 갖게 될텐데 바로 distance가 key가 되지 않고 한번 더 감싸주신 이유가 궁금해요! (fare도 마찬가지로요)

Combi153 replied:
헉 json을 생각하지 않고 코드만 생각하다보니 발생한 실수입니다! 😭 

distance와 fare 가 모두 Integer 타입입니다. 따라서 PathResponse DTO를 생성할 때 헷갈릴 여지가 있다고 생각했습니다. 그래서 distance와 fare를 모두 포장해서 사용하는 방법을 택했습니다.

시간이 지나서 다시 생각해보니, 오버 엔지니어링이었고, json 을 복잡하게 만든 코드가 되었네요 얼른 수정하겠습니다!

seovalue commented on src/main/java/subway/controller/PathController.java:
```diff
@@ -0,0 +1,26 @@
+        this.pathService = pathService;
+    }
+
+    @GetMapping("/shortest")
```
endpoint에 대해 어떻게 생각하시나요? 적절한가요? 페어가 작업해주신 것 같아서 여쭈어요ㅎㅎ

Combi153 replied:
Restful API를 적절히 만족시키는 것은 참 어려운 것 같습니다!

shortest 보다는 다른 명사를 사용하는 것이 흔히 말하는 Rest의 조건을 만족하는 것 같습니다.

/shortest-path 로 수정하겠습니다!

조앤이 느끼시기에 /shortest-path는 괜찮으실까요?

Combi153 replied:
그리고 Restful API에서 GET 메서드에 body를 잘 사용하지 않는 것을 알게 되었습니다. query string으로 요청을 보낼 수 있도록 리팩터링 하겠습니다.

seovalue commented on src/main/java/subway/application/PathService.java:
```diff
@@ -0,0 +1,61 @@
+    public PathResponse findShortestPath(final PathRequest pathRequest) {
+        final SubwayMap subwayMap = new SubwayMap(lineRepository.findLines());
+        final PathFinder pathFinder = new PathFinder(subwayMap);
+        final Station from = stationRepository.findStationById(pathRequest.getFromStationId());
+        final Station to = stationRepository.findStationById(pathRequest.getToStationId());
```
이렇게 `findStationById`를 해서 from, to를 반환하는 쌍이 잦게 호출되는 것 같은데, from/to를 갖는 DTO를 만들어서 그걸 반환하게 해보는 것은 어떨까요?

Combi153 replied:
말씀하신 것처럼 from, to를 한번에 반환하는 경우가 많습니다. 그래서 처음에는 조앤의 의견에 동의하고 DTO를 만들었습니다. 그러나 다시 DTO를 삭제하고 본래의 코드로 돌아갔습니다. DTO를 만들었을 때 오히려 복잡성이 증가하는 것처럼 느껴졌기 때문입니다. 

개인적으로 복잡성이 더 증가했다고 느낀 이유는 from/to를 갖는 DTO의 이름을 짓기 어려웠기 때문입니다.

from/to를 갖는 객체는 현재 Section이 이미 있습니다. 그렇다면 새롭게 만들 DTO는  Section과 완전히 구별되어야 합니다. 다른 개발자들이 보기에 Section과 DTO의 이름이 헷갈리면 안 되기 때문입니다. 한편, 새로운 DTO는 StationRepository와 Service 간에만 사용하는 DTO임이 드러나야 합니다. 이러한 조건을 만족하는 적절한 이름을 떠올리기 어려웠습니다. 가령, StationPair 등의 이름을 사용해보려 했지만, 코드의 가독성이 오히려 떨어지는 것처럼 느껴졌습니다.

결국 DTO를 도입하지 않았습니다. 현재로서는 from, to를 각각 호출하는 것도 괜찮다고 생각합니다.

조앤은 어떻게 생각하시나요?

seovalue replied:
오호! 저도 후추의 의견이 이해가 되어서 공감이 되네요ㅎㅎ Section과의 명확한 차이가 없으며 from, to를 각각 호출하는게 오히려 비즈니스 로직을 더 잘 드러내는 걸로 여겨질 수 있을 것 같아요 👍

seovalue commented on src/test/java/subway/application/LineServiceTest.java:
```diff
@@ -46,7 +52,7 @@ class LineServiceTest {
     void 호선을_저장한다() {
         //given
-        when(lineRepository.save(any(LineEntity.class)))
+        when(lineRepository.save(any(Line.class)))
```
테스트를 정말 꼼꼼히 짜주셨군요 💯 
호선 저장이나 역 추가는 dao 테스트랑도 겹치는 것 같은데 이렇게 중복되는 테스트가 많을 경우엔 추후 유지보수하기도 까다로울 것 같아요. 이런 경우는 어떤 레이어를 테스트하면 모두 커버할 수 있을까요?

Combi153 replied:
호선 저장이나 역 추가 모두 LineService의 기능을 테스트하고 있습니다. 호선 저장은 LineService의 saveLine 메서드를, 역 추가는 addStationToLine 메서드를 테스트 합니다. 따라서 dao 테스트와 중복되지 않는다고 판단했습니다.

만약 레이어 간 테스트하는 동작이 중복된다면, 해당 기능을 갖는 클래스를 테스트하는 것이 맞다고 생각합니다!

혹시 제가 잘못 이해하거나, 잘못 생각한 부분이 있을까요? 조앤의 의견을 말씀해주신다면 감사하겠습니다!

seovalue replied:
호선을 저장한다는 것에 대한 핵심 기능은 dao에서 insert하는 부분이라고 생각이 들어서, 중복이라는 생각이 들었어요! 결론적으로 dao에서 insert만 잘 된다면, `호선 저장`이라는 기능은 잘 동작한다고 여겨져서요! 하지만 후추가 말씀해주신 것처럼 service layer의 테스트가 필요한 것 같아요. insert시 validation 등을 하기 때문에요!ㅎㅎ 둘 다 있어도 좋은 것 같네요 👍

seovalue commented on src/test/java/subway/integration/IntegrationTest.java:
```diff
@@ -3,11 +3,15 @@
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Transactional
+@Sql("classpath:test_data.sql")
```
PathController도 추가되었으니 PathIntegrationTest도 추가해볼까용?

Combi153 replied:
넵 최단 경로 테스트가 빠졌었네요 PathIntegrationTest에 추가하겠습니다!

seovalue reviewed (changes requested):
안녕하세요, 후추ㅎㅎ 조앤이에요
테스트를 정말 꼼꼼하게 짜주셨군요👍 2단계도 잘 수행해주셨네요
몇가지 리뷰 남겼는데, 확인 후 다시 리뷰 요청 주세요~
혹시 3단계도 하시고 싶으시다면 리뷰 먼저 반영 후 해당 PR에 추가로 구현해주셔도 돼요!

Combi153 committed:
refactor: 필요없는 docker 파일 제거

Combi153 committed:
refactor: mapper 추가, entity에서 도메인 변환 로직 제거

Combi153 committed:
refactor: test fixture 수정

Combi153 committed:
test: PathIntegrationTest 추가

Combi153 committed:
refactor: 최단 경로 조회 API 수정

Combi153 committed:
refactor: DistanceResponse, FareResponse 삭제

Combi153 committed:
refactor: 요금 계산 로직 수정

변경 내용 :
- DistanceFarePolicy 추상클래스 도입
- First, Second, LastDistancePolicy 구현체 도입
- Distance subtract 기능 변경
- 기존 요금 계산 로직 삭제

Combi153 committed:
refactor: 구간 수정 메서드 이름 변경

Combi153 commented:
안녕하세요 조앤! 후추입니다 😀 

늦게 재리뷰 요청을 보냅니다.

고백하자면 지하철 미션을 수행하면서 번아웃이 왔습니다.

지쳤다고 느꼈을 때 빠르게 조치를 취했다면 좋았겠지만, 스스로 번아웃이라고 인정하는 것이 왠지 꺼려졌습니다. 자존심을 부렸던 것 같아요. 그렇게 할 일을 꾸역꾸역 억지로 하면서 날이 꽤 흘렀습니다. 그러다 지난주부터는 정말 아무것도 손에 잡히지 않았고, 결국 번아웃을 인정했습니다. 

그후 회복에 집중했습니다. 푹 쉬고 났더니, 다행히 오늘은 꽤나 할 일을 할 수 있었어요. 차차 나아질 것 같은 예감이 듭니다!

그런 이유로 리뷰 요청이 늦어졌습니다. 죄송합니다.

조앤은 번아웃을 겪여보셨나요? 번아웃에서 어떻게 회복하셨는지, 번아웃을 예방하는 방법은 무엇인지 조앤의 경험과 생각이 무척 궁금합니다!

seovalue commented on src/main/java/subway/application/PathService.java:
```diff
@@ -42,7 +40,7 @@ public PathResponse findShortestPath(final PathRequest pathRequest) {
         final Path path = pathFinder.findShortest(from, to);
         final int fare = farePolicy.calculate(path.getDistance());
-        return new PathResponse(toResponses(path.getSectionEdges()), DistanceResponse.of(path.getDistance()), new FareResponse(fare));
+        return new PathResponse(toResponses(path.getSectionEdges()), path.getDistanceValue(), fare);
```
`PathResponse`를 생성할 때 어떤 건 response를 받고, 어떤 것은 객체를 받도록 되어있는데요. 하나로 통일해보면 어떨까요?

seovalue commented on src/main/java/subway/controller/PathController.java:
```diff
@@ -19,8 +19,8 @@ public PathController(final PathService pathService) {
-    public ResponseEntity<PathResponse> findShortestPath(@RequestBody PathRequest pathRequest) {
-        return ResponseEntity.ok(pathService.findShortestPath(pathRequest));
+    @GetMapping("/shortest-path")
+    public ResponseEntity<PathResponse> findShortestPath(@RequestParam Long fromStationId, @RequestParam Long toStationId) {
```
👍 좋네요

seovalue commented on src/main/java/subway/domain/fare/DistanceFarePolicy.java:
```diff
@@ -0,0 +1,10 @@
+package subway.domain.fare;
+
+public abstract class DistanceFarePolicy implements FarePolicy {
```
와우 알잘딱깔센으로 적용해주셨군요 👍  혹시 어떠셨나요? 이렇게 바꿨을 때 좀 더 유지보수에 용이해질 것 같으신가요?ㅎㅎ

seovalue commented on src/main/java/subway/dto/PathResponse.java:
```diff
@@ -5,27 +5,27 @@
     private List<LineAndSectionsResponse> lineAndSectionsResponse;
-    private DistanceResponse distanceResponse;
-    private FareResponse fareResponse;
+    private Integer distance;
+    private Integer fare;
```
👍

seovalue commented on src/main/java/subway/repository/Mapper.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+@Component
+public class Mapper {
```
변환을 담당하는 객체를 만들어주신 이유가 궁금해요!

seovalue commented on .gitignore:
```diff
@@ -30,4 +30,5 @@ out/
-.vscode/
\ No newline at end of file
+.vscode/
+/docker/db/mysql/
```
여기에 남겨볼까요~ ㅋㅋ 마땅한 newLine이 없네요
번아웃을 경험하셨군요! 정말 열심히 하시는 분들이라면 한번쯤은 거쳐가시는 것 같아요. 
저는 특별히 `번아웃이다!` 라고 느끼는 경험을 해 본적이 없어서 조언을 드리기가 어렵지만, 말씀드리고 싶은 부분은, 
번아웃을 경험하신다는 것 자체가 후추가 정말 100%를 우테코에 쏟아붓고 계시기 때문이라고 생각해요. 지금 이 순간 정말 열심히 몰입해서 경험하고 성장하시고 있다는 것을 증명하는 것이니 나쁘게 생각하시지 말고, 좀 더 큰 성장을 위해 잠시 쉬어가는 시간을 가지셨다고 생각하시면 좋을 것 같아요. 상황이 상황인지라 모든 것을 다 놓고 쉴 순 없지만, 전보다는 조금 더 후추 본인을 위한 시간도 가지시며 약간의 여유를 가져보시는 건 어떨까요?ㅎㅎ 그리고 다가올 방학 때엔 정말 푹 쉬어 보시는 것을 추천할게요. 고생많으셨어요 👍  언제나 화이팅이에요.

seovalue reviewed (approved):
안녕하세요 후추~ 조앤이에요ㅎㅎ
리팩터링 잘 해주셨네요! 2단계는 여기서 머지할게요.
남겨주신 질문에 대한 답변도 코멘트로 남겨뒀으니 확인해주세요~
언제나 화이팅이에요! 👍
